src/trust_region.jl

```diff
@@ -36,26 +36,44 @@ function dogleg!{T}(p::Vector{T}, r::Vector{T}, d2::Vector{T}, J::AbstractMatrix
         end
     end
     scale!(p_i, -one(T))
+
     # Test if Gauss-Newton step is within the region
-    if wnorm(d2, p_i) <= delta
-        copy!(p, p_i)
+    if wnorm(d, p_i) <= delta
+        copy!(p, p_i)   # accepts equation 4.13 from N&W for this step
     else
-        p_c = At_mul_B(J, r) # Gradient direction
-        broadcast!(/, p_c, p_c, d2)
-        scale!(p_c, - wnorm(d2, p_c)^2 / sum(abs2,J * p_c)) # Cauchy point
-        if wnorm(d2, p_c) >= delta
+        # For intermediate we will use the output array `p` as a buffer to hold
+        # the gradient. To make it easy to remember which variable that array
+        # is representing we make g an alias to p and use g when we want the
+        # gradient
+
+        # compute g = J'r ./ (d .^ 2)
+        g = p
+        At_mul_B!(g, J, r)
+        broadcast!(/, g, g, d)
+        broadcast!(/, g, g, d)
```

Comment:
Good call. I think we eventually just want `g .= g ./ (d .^ 2)`

I was doing it this way to avoid allocations on 0.5 where the dotted arithmetic operators don't get fused into the anonymous function handed to `broadcast!`